content/blog/2021/08/19/WritingFastConstraintStreamsSecretRecipe.adoc

```diff
@@ -0,0 +1,119 @@
+== Constraint Streams by default
+
+To get you started, OptaPlanner has always brought a https://www.optaplanner.org/download/download.html[wide selection of examples] and more recently https://github.com/kiegroup/optaplanner-quickstarts[quickstarts].
+While quickstarts have always used Constraint Streams and Constraint Streams only, the examples being older in age, they typically bring more scoring implementations, such as https://docs.optaplanner.org/latestFinal/optaplanner-docs/html_single/index.html#droolsScoreCalculation[DRL] or https://docs.optaplanner.org/latestFinal/optaplanner-docs/html_single/index.html#incrementalJavaScoreCalculation[incremental Java].
+To clearly state that Constraint Streams are now the scoring method of choice, we have recently converted all OptaPlanner examples to use Constraint Streams by default.
+That said, we do not intend to deprecate any of the other scoring methods any time soon.
+
+== New constraint collectors
+
+Seeing increased adoption of Constraint Streams, we have been expanding the selection of https://docs.optaplanner.org/latestFinal/optaplanner-docs/html_single/index.html#constraintStreamsGroupingAndCollectors[constraint collectors] available out of the box.
+Recently, we have added the following new constraint collectors:
+
+* `average()` constraint collector allows you to calculate an average of a group of items.
+* `compose()` collector allows you to merge results of several constraint collectors. (For example, the `average()` collector is a composite of `count()` and `sum()`.)
+* `conditionally()` constraint collectors allows you to only delegate to another collector if a given condition is met first.
+
+Especially with the latter two collectors, the expressive power of Constraint Streams has grown significantly.
+
+== Faster constraint collectors
+
+Until recently, constraint collectors `toList()`, `toSet()`, `toSortedSet()`, `toMap()` and `toSortedMap()` have been comparatively slow.
+The reason for that was that, on every incremental change of the inputs, the resulting collections were built all over again.
+We have changed the underlying implementation to be much more friendly to incremental calculation, and the end result is a performance improvement on the order of magnitudes on large enough data sets.
+
+== Experimental constraint collectors
+
+As we were implementing constraint providers for all the various examples, we noticed some constraints (which were already hard to implement in DRL) were impossible with Constraint Streams as it stands.
+Consider the following https://docs.optaplanner.org/latestFinal/optaplanner-docs/html_single/index.html#nurseRostering[Nurse Rostering] https://github.com/kiegroup/optaplanner/blob/8.10.0.Final/optaplanner-examples/src/main/resources/org/optaplanner/examples/nurserostering/solver/nurseRosteringConstraints.drl#L97[DRL-based constraint] to penalize too many consecutive shifts:
+
+[source, drl]
+----
+	rule "insertEmployeeConsecutiveAssignmentStart"
```

Comment:
Tip: if you reduce the indentation by one level, you'll use the horizontal space more efficiently.